submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+## Globalnet 2.0: Enhancements to Globalnet Implementation
+
+## Summary
+
+Currently, Submariner Globalnet implementation in [release 0.8](https://github.com/submariner-io/releases/releases/tag/v0.8.0)
+annotates every Pod, Service and Node in the cluster with a globalIp to support Pod to remote
+Service connectivity in the following manner.
+
+Once a Pod is scheduled and enters Running state, it is annotated with a globalIp. To support
+the connectivity, Globalnet programs the necessary egress rules on the active Gateway Node of
+the cluster where the source IP of the Pod is modified to the corresponding globalIp. This
+translation enables us to uniquely identify the traffic from the Pod, when it reaches the
+destination cluster.
+
+Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
+with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIp.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+1. By default, every cluster will be assigned a single globalIp which will be used as egress IP
+   for any cross-cluster communication.
+2. Provision to allocate a globalIp at the namespace/project level which takes precedence over
```

Comment:
Same here, we should allow a CIDR (or single IP).

Reply:
it's a good point

Reply:
AFAICT a single globalIP could be sufficient at the namespace level. 

Let me explain:
In Linux, ephemeral port range can be seen/configured via `sysctl net.ipv4.ip_local_port_range`. The default range is `net.ipv4.ip_local_port_range = 32768    60999` (i.e., approx 27K). Connection tracking is done based on 5 tuple - i.e., SrcIP, SrcPort, DestIP, DestPort and Protocol. In our use-case we are talking about Egress traffic at namespace level, so SrcIP is globalIP allocated to namespace and SrcPort is ephemeral port range.

In order for ephemeral port exhaustion to happen at the namespace level, if all the Pods in the namespace connect to the same DestIP:DestPort we will run out after 27K connections are established. 

Please note that if pods try to connect to different DestIPs and/or DestPorts, each combination has 27K limit.
So based on this, IMHO a single globalIP should suffice. 

If you think, I'm wrong or missing something, do let me know.

@mkolesnik @mangelajo @aswinsuryan @nyechiel